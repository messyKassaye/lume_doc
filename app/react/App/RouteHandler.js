/** @format */

import { I18NUtils } from 'app/I18N';
import { isClient } from 'app/utils';
import api from 'app/utils/api';
import { RequestParams } from 'app/utils/RequestParams';
import moment from 'moment';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import JSONUtils from 'shared/JSONUtils';

const getLocale = ({ store }) => store.getState().locale;

const setLocale = locale => {
  moment.locale(locale);
  api.locale(locale);
  I18NUtils.saveLocale(locale);
};

class RouteHandler extends Component {
  static async requestState(_requestParams, _state) {
    return new Promise((resolve, _reject) => {
      resolve([]);
    });
  }

  emptyState() {} //eslint-disable-line

  static renderTools() {}

  //eslint-disable-next-line
  isRenderedFromServer() {
    const result = RouteHandler.renderedFromServer;
    RouteHandler.renderedFromServer = false;
    return result;
  }

  constructor(props, context) {
    super(props, context);
    setLocale(getLocale(context));
    if (!this.isRenderedFromServer() && isClient) {
      this.getClientState(this.props);
    }
  }

  componentDidUpdate(prevProps) {
    if (this.urlHasChanged(prevProps)) {
      this.emptyState();
      this.getClientState(this.props);
    }
  }

  async getClientState(props) {
    let query;
    if (props.location) {
      query = JSONUtils.parseNested(props.location.query);
    }

    const { store = { getState: () => {} } } = this.context;

    const headers = {};
    const { lang, ...params } = props.params;
    const requestParams = new RequestParams({ ...query, ...params }, headers);
    const actions = await this.constructor.requestState(requestParams, store.getState());

    actions.forEach(action => {
      store.dispatch(action);
    });
  }

  urlHasChanged(prevProps) {
    const { params: nextParams = {}, routes: nextRoutes = [] } = prevProps;
    const { params, routes } = this.props;

    const sameParams = Object.keys(nextParams).reduce(
      (memo, key) => memo && prevProps.params[key] === params[key],
      true
    );
    const sameAmountOfparams = Object.keys(nextParams).length === Object.keys(params).length;
    const currentPath = routes.reduce((path, r) => path + r.path, '');
    const newPath = nextRoutes.reduce((path, r) => path + r.path, '');
    const samePath = currentPath === newPath;
    return !sameParams || !sameAmountOfparams || !samePath;
  }

  render() {
    return <div>{false}</div>;
  }
}

RouteHandler.renderedFromServer = true;

RouteHandler.defaultProps = {
  params: {},
};

RouteHandler.contextTypes = {
  getInitialData: PropTypes.func,
  isRenderedFromServer: PropTypes.func,
  router: PropTypes.object,
  store: PropTypes.object,
};

RouteHandler.propTypes = {
  params: PropTypes.object,
  routes: PropTypes.array,
  location: PropTypes.object,
};

export default RouteHandler;
