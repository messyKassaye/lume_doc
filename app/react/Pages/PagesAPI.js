import api from 'app/utils/api';

export default {
  get(requestParams) {
    return api.get('pages', requestParams).then(response => response.json);
  },

  getById(requestParams) {
    return api.get('page', requestParams).then(response => response.json);
  },

  save(requestParams) {
    return api.post('pages', requestParams).then(response => response.json);
  },

  delete(requestParams) {
    return api.delete('pages', requestParams).then(response => response.json);
  },
};
