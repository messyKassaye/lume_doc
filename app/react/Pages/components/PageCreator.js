import { Form, Field } from 'react-redux-form';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { MarkDown } from 'app/ReactReduxForms';
import {
  resetPage as resetPageAction,
  savePage as savePageAction,
} from 'app/Pages/actions/pageActions';
import ShowIf from 'app/App/ShowIf';
import { BackButton } from 'app/Layout';
import { Icon } from 'UI';

import validator from './ValidatePage';

export class PageCreator extends Component {
  componentWillUnmount() {
    const { resetPage } = this.props;
    resetPage();
  }

  render() {
    const { formState, page, savePage, savingPage } = this.props;
    const backUrl = '/settings/pages';
    const pageUrl = `/page/${page.data.sharedId}`;

    let nameGroupClass = 'template-name form-group';
    if (
      formState.title &&
      !formState.title.valid &&
      (formState.submitFailed || formState.title.touched)
    ) {
      nameGroupClass += ' has-error';
    }

    return (
      <div className="page-creator">
        <Form model="page.data" onSubmit={savePage} validators={validator()}>
          <div className="panel panel-default">
            <div className="metadataTemplate-heading panel-heading">
              <div className={nameGroupClass}>
                <Field model=".title">
                  <input placeholder="Page name" className="form-control" />
                </Field>
              </div>
            </div>
            <div className="panel-body page-viewer document-viewer">
              <ShowIf if={Boolean(page.data._id)}>
                <div className="alert alert-info">
                  <Icon icon="angle-right" /> {pageUrl}
                  <a
                    target="_blank"
                    rel="noopener noreferrer"
                    href={pageUrl}
                    className="pull-right"
                  >
                    &nbsp;(view page)
                  </a>
                </div>
              </ShowIf>
              <MarkDown htmlOnViewer model=".metadata.content" rows={18} />
              <div className="alert alert-info">
                <Icon icon="info-circle" size="2x" />
                <div className="force-ltr">
                  Use{' '}
                  <a
                    target="_blank"
                    rel="noopener noreferrer"
                    href="https://guides.github.com/features/mastering-markdown/"
                  >
                    Markdown
                  </a>{' '}
                  syntax to create page content
                  <br />
                  You can also embed advanced components like maps, charts and document lists in
                  your page.&nbsp;
                  <a
                    target="_blank"
                    rel="noopener noreferrer"
                    href="https://github.com/huridocs/uwazi/wiki/Components"
                  >
                    Click here
                  </a>{' '}
                  to learn more about the components.
                </div>
              </div>
              <div>
                <div>
                  <span className="form-group-label">Page Javascript</span>
                </div>
                <div className="alert alert-warning">
                  <Icon icon="exclamation-triangle" size="2x" />
                  <div className="force-ltr">
                    With great power comes great responsibility!
                    <br />
                    <br />
                    This area allows you to append custom Javascript to the page. This opens up a
                    new universe of possibilities.
                    <br />
                    It could also very easily break the app. Only write code here if you know
                    exactly what you are doing.
                  </div>
                </div>
                <Field model=".metadata.script">
                  <textarea
                    placeholder="// Javascript - With great power comes great responsibility!"
                    className="form-control"
                    rows={12}
                  />
                </Field>
              </div>
            </div>
          </div>
          <div className="settings-footer">
            <BackButton to={backUrl} />
            <button type="submit" className="btn btn-success save-template" disabled={!!savingPage}>
              <Icon icon="save" />
              <span className="btn-label">Save</span>
            </button>
          </div>
        </Form>
      </div>
    );
  }
}

PageCreator.propTypes = {
  resetPage: PropTypes.func,
  savePage: PropTypes.func,
  page: PropTypes.object,
  formState: PropTypes.object,
  savingPage: PropTypes.bool,
};

function mapStateToProps({ page }) {
  return { page, formState: page.formState, savingPage: page.uiState.get('savingPage') };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ resetPage: resetPageAction, savePage: savePageAction }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PageCreator);
