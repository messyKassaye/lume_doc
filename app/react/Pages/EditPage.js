import React from 'react';
import { actions as formActions } from 'react-redux-form';

import RouteHandler from 'app/App/RouteHandler';

import PageCreator from 'app/Pages/components/PageCreator';
import pagesAPI from './PagesAPI';

export default class EditPage extends RouteHandler {
  static async requestState(requestParams) {
    const [page] = await pagesAPI.get(requestParams);
    return [formActions.load('page.data', page)];
  }

  componentDidMount() {
    this.context.store.dispatch(formActions.reset('page.data'));
  }

  render() {
    return <PageCreator />;
  }
}
