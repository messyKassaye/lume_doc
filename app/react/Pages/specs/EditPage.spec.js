import React from 'react';
import { shallow } from 'enzyme';

import PagesAPI from 'app/Pages/PagesAPI';
import PageCreator from 'app/Pages/components/PageCreator';
import RouteHandler from 'app/App/RouteHandler';
import EditPage from '../EditPage';

describe('EditPage', () => {
  const page = { _id: 'abc2', title: 'Page 1' };
  let component;
  let context;

  beforeEach(() => {
    spyOn(PagesAPI, 'get').and.returnValue(Promise.resolve([page]));
    RouteHandler.renderedFromServer = true;
    context = { store: { getState: () => ({}), dispatch: jasmine.createSpy('dispatch') } };
    component = shallow(<EditPage />, { context });
  });

  it('should render a PageCreator', () => {
    expect(component.find(PageCreator).length).toBe(1);
  });

  describe('static requestState()', () => {
    it('should request page being edited', async () => {
      const data = { pageId: 'abc2' };
      const request = { data };
      const actions = await EditPage.requestState(request);

      expect(PagesAPI.get).toHaveBeenCalledWith(request);
      expect(actions).toMatchSnapshot();
    });
  });
});
