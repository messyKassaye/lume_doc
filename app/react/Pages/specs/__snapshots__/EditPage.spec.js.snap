// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPage static requestState() should request page being edited 1`] = `
Array [
  Object {
    "external": true,
    "load": true,
    "model": "page.data",
    "multi": false,
    "silent": true,
    "type": "rrf/change",
    "value": Object {
      "_id": "abc2",
      "title": "Page 1",
    },
  },
]
`;
